src/box/lua/net_box.c

```diff
@@ -123,6 +123,7 @@ enum netbox_state {
 	NETBOX_ERROR             = 4,
 	NETBOX_ERROR_RECONNECT   = 5,
 	NETBOX_CLOSED            = 6,
```

Comment:
This is intentional - `conn:call` waits for the connection to become active or switch to a final state ('error' or 'close'). 'graceful_shutdown` isn't considered a final state, because if `reconnect_after` option is set, the connection may be reestablished.

However, you're the second person who's confused by this behavior so I guess we should make 'graceful_shutdown' a final state if `reconnect_after` is not set. Fixed. The diff is below.

<details><summary>diff</summary>

```diff
diff --git a/src/box/lua/net_box.lua b/src/box/lua/net_box.lua
index cc6999b4cc4c..bb593bfcf4cf 100644
--- a/src/box/lua/net_box.lua
+++ b/src/box/lua/net_box.lua
@@ -813,6 +813,8 @@ function remote_methods:wait_state(state, timeout)
     -- FYI: [] on a string is valid
     repeat until self.state == state or state[self.state] or
                  self.state == 'closed' or self.state == 'error' or
+                 (not self.opts.reconnect_after and
+                  self.state == 'graceful_shutdown') or
                  not self._state_cond:wait(max(0, deadline - fiber_clock()))
     return self.state == state or state[self.state] or false
 end
diff --git a/test/box-luatest/graceful_shutdown_test.lua b/test/box-luatest/graceful_shutdown_test.lua
index 660a62cff49b..da510f7f2ceb 100644
--- a/test/box-luatest/graceful_shutdown_test.lua
+++ b/test/box-luatest/graceful_shutdown_test.lua
@@ -148,7 +148,7 @@ g.test_new_requests_not_allowed = function()
     t.assert(conn:wait_state('graceful_shutdown'))
     t.assert_error_msg_content_equals(
         'Connection is not established, state is "graceful_shutdown"',
-        conn.eval, conn, 'return true', {}, {timeout = 0.01})
+        conn.eval, conn, 'return true')
     local conn2 = net.connect(g.server.net_box_uri)
     t.assert_not(conn2:ping())
     t.assert(conn2.state, 'error')
```

</details>